questionnaires/forms.py

```diff
@@ -169,31 +177,51 @@ def add_stream_field_error(self, position, field, message):
     @property
     def clean_errors(self):
         if self._clean_errors.keys():
-            params = {
-                key: ErrorList(
-                    [ValidationError('Error in form', params=value)]
-                )
-                for key, value in self._clean_errors.items()
-                if isinstance(key, int)
-            }
             errors = {
                 key: ValidationError(value)
                 for key, value in self._clean_errors.items()
                 if isinstance(key, str)
             }
-            errors.update({
-                'skip_logic': ErrorList([ValidationError(
-                    'Skip Logic Error',
-                    params=params,
-                )])
-            })
-            return errors
+
+            blocks_errors = {}
+            for block_index, errors_dict in self._clean_errors.items():
+                if isinstance(block_index, int):
+                    block_errors = {}
+                    for field, errors_list in errors_dict.items():
+                        block_errors[field] = ErrorList([ValidationError(message) for message in errors_list])
+                    else:
+                        blocks_errors[block_index] = ErrorList([StructBlockValidationError(block_errors=block_errors)])
+            else:
+                stream_block_errors = StreamBlockValidationError(block_errors=blocks_errors, non_block_errors=ErrorList())
+                errors.update({
+                    "skip_logic": ErrorList([stream_block_errors]),
+                })
+
+            return ErrorDict(errors)
 
     def add_form_field_error(self, field, message):
         if field not in self._clean_errors:
             self._clean_errors[field] = list()
         self._clean_errors[field].append(message)
 
+    def should_be_multi_step(self):
+        cleaned_data = super().clean()
+        question_data = {}
+
+        for form in self.formsets[self.form_field_name]:
+            form.is_valid()
+            question_data[form.cleaned_data['ORDER']] = form
+
+        for i, form in question_data.items():
+            if form.is_valid():
+                data = form.cleaned_data
+                for logic in data['skip_logic']:
+                    if (logic.value['skip_logic'] in [SkipState.QUESTION, SkipState.END]
+                            and not cleaned_data['multi_step']):
+                        return True
+
+        return False
```

Comment:
I don't feel comfortable calling `clean()` and `is_valid()` again because I don't think it is intended for it to be called more than once, and it doesn't seem necessary to do so here. I don't mind looping through existing, clean data, but I did want to avoid re-cleaning the data. I think this should be modified to something like:

```
def should_be_multi_step(cleaned_data, question_data):
    ...
```

That is, the method should be a class method that we pass the cleaned data to. This should simplify the method quite a bit.

In addition, please make sure to clean things up a bit. For example:

```
for i, form in question_data.items():
    ...
````

The key/index `i` is not used so we should loop through values rather than items.